ggshield/cmd/iac/scan/diff.py

```diff
@@ -57,57 +63,103 @@ def scan_diff_cmd(
+    :param directory: path to the location we want to scan. Defaults to the current
+    directory.
+    :param previous_ref: git reference to the state of reference for the analysis
+    :param include_staged: bool whether or not we want to consider the staged files
+    only when the current reference is set to None.
+    :param current_ref: optional git reference to the current state, defaults to None.
+    When set to None, the current state is the indexed files currently on disk.
+    :return: IacDiffScanResult if the scan was performed; IaCSkipScanResult if the scan
+    was skipped (i.e. no IaC files were detected or changed between the two references)
+    """
     config = ctx.obj["config"]
     client = ctx.obj["client"]
     exclusion_regexes = ctx.obj["exclusion_regexes"]
 
     verbose = config.user_config.verbose if config and config.user_config else False
     if verbose:
-        display_info(f"> Scanned files in reference {ref}")
-        filepaths = filter_iac_filepaths(directory, get_git_filepaths(directory, ref))
-        for filepath in filepaths:
-            display_info(f"- {click.format_filename(filepath)}")
-        display_info("")
-
-    current_ref = INDEX_REF if include_staged else "HEAD"
+        if previous_ref is None:
+            display_info(
+                "> No file to scan in reference. This might be a new repository."
+            )
+        else:
+            display_info(f"> Scanned files in reference {previous_ref}")
+            filepaths = filter_iac_filepaths(
+                directory, get_git_filepaths(directory, previous_ref)
+            )
+            for filepath in filepaths:
+                display_info(f"- {click.format_filename(filepath)}")
+            display_info("")
+
+    if current_ref is None:
+        current_ref = INDEX_REF if include_staged else "HEAD"
     if verbose:
         if include_staged:
             display_info("> Scanned files in current state (staged)")
         else:
             display_info("> Scanned files in current state")
         filepaths = filter_iac_filepaths(
-            directory, get_git_filepaths(directory, current_ref)
+            directory, get_git_filepaths(directory=directory, ref=current_ref)
         )
         for filepath in filepaths:
             display_info(f"- {click.format_filename(filepath)}")
 
+    modified_iac_files = []
+
     # Check if IaC files were created, deleted or modified
-    files_status = get_diff_files_status(
-        wd=str(directory), ref=ref, staged=include_staged, similarity=100
-    )
-    modified_modes = [Filemode.NEW, Filemode.DELETE, Filemode.MODIFY]
-    modified_iac_files = [
-        file
-        for file, mode in files_status.items()
-        if mode in modified_modes
-        and not is_filepath_excluded(str(file), exclusion_regexes)
-        and is_iac_file_path(file)
-    ]
+    if previous_ref is None:
+        # This means we are scanning all commits up to now.
```

Comment:
Here the comment says "we are scanning all commits up to now".
However we previously displayed "No file to scan in reference. This might be a new repository" for the previous_ref is None case.

What is the real use case of previous_ref is none ?

Reply:
`previous_ref` is None when the pre-receive command failed to find a root for a newly pushed branch.

Here, `None` is interpreted as the global root for the git tree. I wish we could rely on `EMPTY_TREE` or `EMPTY_SHA`, but it is dependent on whether people pushed an empty commit when creating the repo.

For pre-receive, setting `previous_ref` to `None` is not exactly equivalent to performing a `scan all` : in the current implementation, scan all requires a directory and file paths. In the pre-receive context, we only have access to git objects. `scan diff` handles that properly, hence the `previous_ref=None`.

Reply:
Ok, clear for me. Thanks for the explications